api/lib/opentelemetry/resources/resource.rb

```diff
@@ -9,18 +9,41 @@ module Resources
     # Resource represents a resource, which captures identifying information about the entities
     # for which telemetry (metrics or traces) is reported.
     class Resource
+      # Returns the labels for this {Resource}
+      #
+      # @return [Hash<String, String>]
       attr_reader :labels
 
-      def initialize(**kvs)
+      # Returns a newly created {Resource} with the specified labels
+      #
+      # @param [Hash<String, String>] kvs Hash of key-value pairs to be used
+      #   as labels for this resource
+      # @return [Resource]
+      def initialize(kvs = {})
         # TODO: how defensive should we be here?
-        @labels = kvs.to_h { |k, v| [k.to_s.clone.freeze, v.to_s.clone.freeze] }.freeze
+        @labels = kvs.each_with_object({}) do |(k, v), memo|
+          memo[-k] = -v
+        end
       end
 
-      # TODO: Already set labels MUST NOT be overwritten unless they are empty string.
+      # Returns a new, merged {Resource} by merging the current {Resource} with
+      # the other {Resource}. In case of a collision, the current {Resource}
+      # takes precedence
+      #
+      # @param [Resource] other The other resource to merge
+      # @return [Resource] A new resource formed by merging the current resource
+      #   with other
       def merge(other)
         raise ArgumentError unless other.is_a?(Resource)
 
-        self.class.new(labels.merge(other.labels))
+        merged_labels = \
+          other.labels.each_with_object(labels.dup) do |(k, v), memo|
+            next if (current = memo[k]) && !current.empty?
+
+            memo[k] = v
+          end
+
+        self.class.new(merged_labels)
```

Comment:
It's unfortunate that we dup the hash here. If we adopted the factory method approach (#25), we could make the constructor private and more flexible.

Reply:
I went the factory method route which spares us the additional hash allocation and type checking. Users will instantiate a resource with `Resource.create` which will type check and freeze labels on the incoming hash. The constructor is private and expects that it receives a frozen hash with frozen string key-value pairs.

Reply:
In regards to #25, we can talk about caching, what that would look like, and whether or not it would be useful.